src/tty/unix.rs

```diff
@@ -126,13 +128,305 @@ impl Read for StdinRaw {
+// they pretty much are)
+type EscapeSeq = smallvec::SmallVec<[u8; 12]>;
+
+/// Basically a dictionary from escape sequences to keys. It can also tell us if
+/// a sequence cannot possibly result in
+struct EscapeBindings {
+    bindings: Trie<EscapeSeq, KeyPress>
+}
+
+impl EscapeBindings {
+    pub fn common_unix() -> Self {
+        let mut res = Self {
+            bindings: Trie::new()
+        };
+        res.add_common_unix();
+        res
+    }
+
+    fn is_prefix<B: ?Sized + AsRef<[u8]>>(&self, invalid_sequence: &B) -> bool {
+        !self.bindings.subtrie(invalid_sequence.as_ref()).is_empty()
+    }
+
+    // Note: the overwrite flag is always false during initialization. It means
+    // we panic if there's a dupe or if one escape is the prefix of another. and
+    // just here for the hard-coded binding list, that we don't accidentally
+    // break various bindings when stuff inevitably gets added to the default
+    // list. If we implement populating the binding set dynamically (via
+    // terminfo, a config file, ...) then we should allow overwriting.
+    //
+    // Also this takes string input only because it debug formats much nicer. As
+    // I mentioned above, escape sequences aren't generally utf8-encoded.
+    fn bind(&mut self, binding: impl Into<KeyPress>, seq_str: &str, overwrite: bool) {
+        let binding = binding.into();
+        let seq = seq_str.as_bytes();
+        assert!(!seq.is_empty());
+        assert!(overwrite || !self.is_prefix(seq), "{:?}", seq_str);
+        assert!(
+            overwrite || !self.bindings.contains_key(seq),
+            "{:?}",
+            seq_str
+        );
+        if seq.len() > 1 && !overwrite {
+            // Check if the thing we're adding is already blocked because it
+            // has a prefix which already exists in the trie
+            for i in 1..seq.len() {
+                assert!(
+                    !self.bindings.contains_key(&seq[..i]),
+                    "{:?} is prefix of {:?}: {:?}",
+                    &seq_str[..i],
+                    seq_str,
+                    binding
+                );
+            }
+        }
+        let existed = self.bindings.insert(seq.into(), binding);
+        if !overwrite {
+            assert_eq!(existed, None, "{:?} {:?}", seq_str, binding);
+        }
+    }
+
+    // Ideally some of this would be read out of terminfo, but... that's a pain
+    // and terminfo doesn't have entries for everything, and sometime's it's
+    // just wrong anyway (*cough* iTerm *cough*). So instead we just add
+    // more-or-less everything we care about to the set of bindings. That said,
+    // this is actually fine. These are the strings that are coming out of the
+    // terminal, so long as two terminals don't use the same strings to mean
+    // different things, it won't be a problem that we have so many items in our list.
+    fn add_common_unix(&mut self) {
+        // Ansi, vt220+, xterm, ...
```

Comment:
To be checked with https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/prompt_toolkit/input/ansi_escape_sequences.py#L68-L312

Reply:
I hadn't seen that, thanks.

Taking a quick look, it seems like they have a few things we're missing, but for the stuff we do support I think we are doing better than them (they don't seem to handle some of the ways iTerm / Terminal.app are configured, as well as some of the xterm sequences we have).

- Our code doesn't handle function keys beyond F12. They go to F24, in theory terminals support up to F64 but... I've never seen a keyboard with more than 12, and the old code stopped at F12 I think.
- They support some cool stuff like scrolling that we should add.
- They support some `mintty`-specific escape sequences, which I wasn't aware of, and would be probably easy to add.

These seem like good things to add, but none of them (AFAIK?) were supported prior to my changes either, so at least they aren't regressions.

That said, I think we should definitely ensure we aren't missing anything, and that page looks good for that. I'm glad to see they have a similar approach of "just listen for everything" instead of following terminfo, which was wrong in many cases I found.

If you want to check my work further, I put this set together using a combination of:

1. Installing a fully updated `terminfo` for various terminals people seem to use and ensuring we support the sequences those advertise in their terminfo (both according to asking around and playing with the https://pkgstats.archlinux.de/ api some).
    - Often the actual terminfo for their default `TERM` variable is lacking, but a combination of seeing if the [terminfo source](https://raw.githubusercontent.com/mirror/ncurses/47d2fb4537d9ad5bb14f4810561a327930ca4280/misc/terminfo.src)
    - That said, this turned out to be wrong or incomplete in a lot of cases, so I went with just listening for all the escapes at once.

2. Reading the XTerm control sequence documentation: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html. This helped a lot, since xterm is at this point practically the standard.

3. A lot of manual testing by installing various terminals, firing up `fish_key_reader -c` (Extremely helpful for this, even if you don't use fish -- prints out exactly what bytes each sequence sends), and seeing what sequences are written in response to different actions.

4. And then leaving in any sequence that the old code handled, even if I couldn't find a source for it, so long as it would be harmless (which they all seemed to be.

Reply:
> Our code doesn't handle function keys beyond F12. They go to F24, in theory terminals support up to F64 but... I've never seen a keyboard with more than 12, and the old code stopped at F12 I think.

You can use `tmux send-keys` to send the higher function keys if you are using tmux.